google/cloud/spanner/result_set.h

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_RESULT_SET_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_RESULT_SET_H_
+
+#include "google/cloud/spanner/internal/time.h"
+#include "google/cloud/spanner/row_parser.h"
+#include "google/cloud/spanner/timestamp.h"
+#include "google/cloud/optional.h"
+#include <google/spanner/v1/spanner.pb.h>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+namespace internal {
+class ResultSetValueSource {
+ public:
+  virtual ~ResultSetValueSource() = default;
+  // Returns OK Status with no Value to indicate end-of-stream.
+  virtual StatusOr<optional<Value>> GetNextValue() = 0;
+  virtual optional<google::spanner::v1::ResultSetMetadata> GetMetadata() = 0;
+  virtual optional<google::spanner::v1::ResultSetStats> GetStats() = 0;
+};
+}  // namespace internal
+
+/**
+ * ResultSet is returned from "query" operations (`Read`, `ExecuteSql`).
+ */
+class ResultSet {
+ public:
+  ResultSet() = default;
+  explicit ResultSet(std::unique_ptr<internal::ResultSetValueSource> source)
+      : source_(std::move(source)) {}
+
+  // This class is moveable but not copyable.
+  ResultSet(ResultSet&&) = default;
+  ResultSet& operator=(ResultSet&&) = default;
+
+  /**
+   * Return a `RowParser` which can be used to iterate through the results.
+   */
+  template <typename... Ts>
+  RowParser<Ts...> rows() {
```

Comment:
Can we have multiple `rows()` iterating over the same `ResultSet`?  If not, maybe this should be a `&&` overload?

We should also consider using a `shared_ptr<>` for the `source_` member (without changing the constructor) so the `RowParser` functor and the `ResultSet` can share the source and have independent lifetimes.

Reply:
After thinking about this some, I think I'm now of the opinion that a clear comment/warning on this function may be the  best way forward. Here's my thinking.

1. There's nothing inherently wrong with calling `rows()` multiple times. A caller could call `rs.rows<...>()`, iterate and consume the first, say, 10 rows, then do something else, then call `rs.rows<...>()` again, and iterate the next 10, and so on. Of course, the caller *could* also do this by saving the first `RowParser` that was returned, but I'm trying to point out that there's nothing fundamentally wrong with calling `rows()` multiple times. I think this would naturally work.

2. The *problem* would arise if a user got multiple `RowParsers` (either from multiple calls to `rows()` or by copying a `RowParser`), and expected that iterating both would produce consistent results. Clearly, that would not work because they're all consuming a single stream of incoming values. But I don't think it's really a problem with `RowParser` nor `ResultSet`, not `ResultSet::rows()`. I think it's just an "invalid usage" problem.

3. Of course, we want to try to make our APIs easy to use correctly and hard to use incorrectly, but I think this might be a case where clear documentation about what users *should* and *should not* do is reasonable.

But I could easily be convinced of some other option :-)

So, uh, thoughts?